subsys/bootloader/bl_crypto/bl_crypto_cc310_hash.c

```diff
@@ -7,94 +7,200 @@
+		ret = nrf_cc310_bl_hash_sha256_update(ctx, source, chunk_len);
 
-			if (nrf_cc310_bl_hash_sha256_update(
-				    &context, (u8_t *)ram_buffer,
-				    block_len) != CRYS_OK) {
-				return false;
-			}
+		if (ret != CRYS_OK) return ret;
 
-			remaining_copy_len -= block_len;
-			block_len = MIN(remaining_copy_len, MAX_CHUNK_LEN);
+		remaining_copy_len -= chunk_len;
+		chunk_len = MIN(remaining_copy_len, max_chunk_len);
+	}
+
+	return CRYS_OK;
+}
+
+
+static inline int hash_blocks_stack(nrf_cc310_bl_hash_context_sha256_t *const ctx,
+		const u8_t *data, u32_t data_len, u32_t chunk_len)
+{
+	u32_t stack_buffer[chunk_len/4];
+	return hash_blocks(ctx, data, data_len, chunk_len, stack_buffer);
+}
+
+/* Base implementation with 'external' parameter. */
+static int sha256_update(nrf_cc310_bl_hash_context_sha256_t *const ctx,
+		const u8_t *data, u32_t data_len, bool external)
+{
+	if ((u32_t)data < CONFIG_SRAM_BASE_ADDRESS) {
+		/* Copy to RAM buffer, then hash. */
+		/* Cryptocell has DMA access to RAM only */
+		if (external) {
+			return hash_blocks_stack(ctx, data, data_len, CHUNK_LEN_STACK);
+		} else {
+			return hash_blocks(ctx, data, data_len, MAX_CHUNK_LEN, ram_buffer);
 		}
 	} else {
-		for (u32_t i = 0; i < data_len; i += MAX_CHUNK_LEN) {
-			if (nrf_cc310_bl_hash_sha256_update(
-				    &context, &data[i],
-				    MIN(data_len - i, MAX_CHUNK_LEN)) !=
-			    CRYS_OK) {
-				return false;
-			}
-		}
+		return hash_blocks(ctx, data, data_len, MAX_CHUNK_LEN, NULL);
 	}
+}
 
-	if (nrf_cc310_bl_hash_sha256_finalize(
-		    &context, (nrf_cc310_bl_hash_digest_sha256_t *)out_hash) !=
-	    CRYS_OK) {
-		return false;
+int bl_sha256_update(nrf_cc310_bl_hash_context_sha256_t *ctx,
+		const u8_t * data, u32_t data_len)
+{
+	int retval = sha256_update(ctx, data, data_len, true);
+	switch (retval) {
+	case CRYS_HASH_INVALID_USER_CONTEXT_POINTER_ERROR:
+	case CRYS_HASH_USER_CONTEXT_CORRUPTED_ERROR:
+		return -EINVAL;
+	case CRYS_HASH_LAST_BLOCK_ALREADY_PROCESSED_ERROR:
+		return -ENOTSUP;
+	default:
+		return retval;
 	}
+}
 
-	return true;
+int bl_sha256_finalize(nrf_cc310_bl_hash_context_sha256_t *ctx, u8_t *output)
+{
+	int retval = nrf_cc310_bl_hash_sha256_finalize(ctx,
+			(nrf_cc310_bl_hash_digest_sha256_t *) output);
+	cc310_bl_backend_disable();
+
+	switch (retval) {
+	case CRYS_HASH_INVALID_USER_CONTEXT_POINTER_ERROR:
+	case CRYS_HASH_USER_CONTEXT_CORRUPTED_ERROR:
+	case CRYS_HASH_INVALID_RESULT_BUFFER_POINTER_ERROR:
+		return -EINVAL;
+	default:
+		return retval;
+	}
+}
+
+bool bl_sha256_verify(nrf_cc310_bl_hash_context_sha256_t *ctx,
+		const u8_t *expected)
+{
+	nrf_cc310_bl_hash_digest_sha256_t hash;
+	int retval = bl_sha256_finalize(ctx, hash);
+	if ((retval == 0) && (memcmp(expected, hash,
+			sizeof(nrf_cc310_bl_hash_digest_sha256_t)) == 0)) {
+		return true;
+	}
+	return false;
 }
 
-bool get_hash(u8_t *hash, const u8_t *data, u32_t data_len)
+int bl_sha256_abort(nrf_cc310_bl_hash_context_sha256_t *ctx)
```

Comment:
What is the final goal of this function?

Reply:
It's part of the ARM PSA crypto API.

Reply:
Okei, nice job keeping it future compatible :)

Reply:
Come to think of it, with versioning of ABIs, it might be better to leave it out as long as it's not implemented, and rather increment the version when it's actually implemented.

Reply:
What do you say @frkv ?

Reply:
Abort is only practical in a multi-threaded operation and/or in a scenario where this is accessed from an NSPE. I don't know how far this "shared interface" is going to be extended. If this is exposed in a non-secure domain then the abort makes sense, but until that point the API is probably a bit superfluous

Reply:
I removed `abort` now. I think not including it will make things less confusing.